test/test_custom_op_testing.py

```diff
@@ -0,0 +1,327 @@
+from torch.testing._internal.optests.compile_check import operator_compile_check
+from torch.testing._internal.custom_op_db import custom_op_db
+from torch._custom_op.impl import custom_op
+
+class TestCustomOpTesting(TestCase):
+    def setUp(self):
+        self.test_ns = '_test_custom_op'
+        self.libraries = []
+
+    def tearDown(self):
+        import torch._custom_op
+        keys = list(torch._custom_op.impl.global_registry.keys())
+        for key in keys:
+            if not key.startswith(f'{self.test_ns}::'):
+                continue
+            torch._custom_op.impl.global_registry[key]._destroy()
+        if hasattr(torch.ops, self.test_ns):
+            del torch.ops._test_custom_op
+        for lib in self.libraries:
+            del lib.m
+        del self.libraries
+
+    def ns(self):
+        return getattr(torch.ops, self.test_ns)
+
+    def lib(self):
+        result = torch.library.Library(self.test_ns, 'FRAGMENT')
+        self.libraries.append(result)
+        return result
+
+    def test_incorrect_schema_mutation(self, device):
+        lib = self.lib()
+        lib.define("foo(Tensor x) -> Tensor")
+        op = self.ns().foo.default
+
+        class Foo(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                guard = torch._C._AutoDispatchBelowAutograd()
+                try:
+                    return op(x)
+                finally:
+                    del guard
+
+            @staticmethod
+            def backward(ctx, gx):
+                return gx
+
+        def foo_impl(x):
+            x.sin_()
+            return x.clone()
+
+        lib.impl("foo", Foo.apply, "Autograd")
+        lib.impl("foo", foo_impl, "CPU")
+        lib.impl("foo", foo_impl, "CUDA")
+
+        def f(x):
+            x = x.clone()
+            v = x.view_as(x)
+            y = op(v)
+            return x
+
+        x = torch.tensor(3.14159 / 3, requires_grad=True, device=device)
+        with self.assertRaisesRegex(
+                RuntimeError,
+                'Argument x is not defined as mutable but was mutated'):
+            operator_compile_check(f, (x,), {})
+
+    def test_incorrect_schema_view(self, device):
+        lib = self.lib()
+        lib.define("foo(Tensor x) -> Tensor")
+        op = self.ns().foo.default
+
+        class Foo(torch.autograd.Function):
+            @staticmethod
+            def forward(ctx, x):
+                # Emulate AutoDispatchBelowADInplaceOrView, which is not bound into python
+                with torch._C._AutoDispatchBelowAutograd():
+                    with torch._C._ExcludeDispatchKeyGuard(torch._C.DispatchKeySet(torch._C.DispatchKey.ADInplaceOrView)):
+                        return op(x)
+
+            @staticmethod
+            def backward(ctx, gx):
+                return gx
+
+        def foo_impl(x):
+            return x.view_as(x)
+
+        def foo_meta(x):
+            return x.view_as(x)
```

Comment:
Just for my understanding, if everything else is correct, does it matter whether the meta has the view-ness as the impl?

Reply:
yep! Especially for functionalization - we rely on the meta functions when we're tracing with functionalization, so functionalization can know what aliases to replay any mutations on.

Reply:
I see, thanks!